File: backend/app/schemas/phd_plan.py
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import datetime, date
from app.models.phd_plan import PhDPlanStatus, VenueType, VenueRating, PaperPlanStatus, ApprovalAction


class PlannedPaperBase(BaseModel):
    paper_number: int = Field(..., ge=1, le=3)
    title: str = Field(..., min_length=5, max_length=500)
    research_question: str = Field(..., min_length=10)
    methodology: Optional[str] = None
    expected_contribution: Optional[str] = None
    target_venue: Optional[str] = Field(None, max_length=300)
    venue_type: Optional[VenueType] = None
    venue_rating: VenueRating
    target_completion_date: date

    @validator('venue_rating')
    def validate_venue_rating(cls, v):
        if v == VenueRating.C:
            raise ValueError("Papers must target venues rated B or higher")
        return v


class PlannedPaperResponse(PlannedPaperBase):
    id: int
    status: PaperPlanStatus
    research_project_id: Optional[int] = None

    class Config:
        from_attributes = True


class PhDPlanBase(BaseModel):
    research_topic: str = Field(..., min_length=5, max_length=500)
    research_question: str = Field(..., min_length=20)
    research_field: Optional[str] = Field(None, max_length=200)
    expected_duration_years: int = Field(4, ge=3, le=7)
    proposal_text: Optional[str] = None
    expose_text: Optional[str] = None


class PhDPlanUpdate(PhDPlanBase):
    papers: Optional[List[PlannedPaperBase]] = None
    change_reason: Optional[str] = None

    @validator('papers')
    def validate_papers(cls, v):
        if v and len(v) != 3:
            raise ValueError("Exactly 3 papers must be planned")
        return v


class PhDPlanSubmit(BaseModel):
    pass  # No additional fields needed for submission


class PhDPlanApproval(BaseModel):
    comment: str = Field(..., min_length=10)


class PhDPlanResponse(BaseModel):
    id: int
    student_id: int
    research_topic: str  # Allow empty for drafts
    research_question: str  # Allow empty for drafts
    research_field: Optional[str] = None
    expected_duration_years: int
    proposal_text: Optional[str] = None
    expose_text: Optional[str] = None
    status: PhDPlanStatus
    current_version: int
    submitted_at: Optional[datetime] = None
    approved_at: Optional[datetime] = None
    approved_by_id: Optional[int] = None
    created_at: datetime
    updated_at: Optional[datetime] = None
    papers: List[PlannedPaperResponse] = []
    proposal_document_id: Optional[int] = None
    expose_document_id: Optional[int] = None

    class Config:
        from_attributes = True


class PhDPlanVersionResponse(BaseModel):
    id: int
    phd_plan_id: int
    version_number: int
    data_snapshot: Dict[str, Any]
    changes_made: Optional[Dict[str, Any]] = None
    created_by_id: int
    created_at: datetime
    change_reason: Optional[str] = None

    class Config:
        from_attributes = True


class PhDPlanApprovalResponse(BaseModel):
    id: int
    phd_plan_id: int
    action: ApprovalAction
    comment: str
    reviewer_id: int
    reviewed_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/types/phd-plan.ts
export enum PhDPlanStatus {
  DRAFT = 'draft',
  SUBMITTED = 'submitted',
  UNDER_REVIEW = 'under_review',
  REVISION_REQUESTED = 'revision_requested',
  APPROVED = 'approved'
}

export enum VenueType {
  CONFERENCE = 'conference',
  JOURNAL = 'journal'
}

export enum VenueRating {
  A_STAR = 'A*',
  A = 'A',
  B = 'B',
  C = 'C'
}

export enum PaperPlanStatus {
  PLANNED = 'planned',
  IN_PROGRESS = 'in_progress',
  SUBMITTED = 'submitted',
  PUBLISHED = 'published'
}

export enum ApprovalAction {
  APPROVE = 'approve',
  REQUEST_REVISION = 'request_revision',
  COMMENT = 'comment'
}

export interface PlannedPaper {
  id?: number;
  paper_number: number;
  title: string;
  research_question: string;
  methodology?: string;
  expected_contribution?: string;
  target_venue?: string;
  venue_type?: VenueType;
  venue_rating: VenueRating;
  target_completion_date: string;
  status?: PaperPlanStatus;
  research_project_id?: number;
}

export interface PhDPlan {
  id: number;
  student_id: number;
  research_topic: string;
  research_question: string;
  research_field?: string;
  expected_duration_years: number;
  proposal_text?: string;
  expose_text?: string;
  status: PhDPlanStatus;
  current_version: number;
  submitted_at?: string;
  approved_at?: string;
  approved_by_id?: number;
  created_at: string;
  updated_at?: string;
  papers: PlannedPaper[];
  proposal_document_id?: number;
  expose_document_id?: number;
}

export interface PhDPlanUpdate {
  research_topic?: string;
  research_question?: string;
  research_field?: string;
  expected_duration_years?: number;
  proposal_text?: string;
  expose_text?: string;
  papers?: PlannedPaper[];
  change_reason?: string;
}

export interface PhDPlanApproval {
  comment: string;
}

export interface PhDPlanVersion {
  id: number;
  phd_plan_id: number;
  version_number: number;
  data_snapshot: any;
  changes_made?: any;
  created_by_id: number;
  created_at: string;
  change_reason?: string;
}

export interface PhDPlanApprovalRecord {
  id: number;
  phd_plan_id: number;
  action: ApprovalAction;
  comment: string;
  reviewer_id: number;
  reviewed_at: string;
}